### PR TITLE
fix: SidecarSet Expectations Leakage Bug

### PR DESCRIPTION
**Pull request by wangwu50**
fix sidecarSet expectations leakage bug when delete pod alfter update sidecarset

### CODE REVIEW

**Comment by wangwu50** [comment]
fixes #1293

File: pkg/controller/sidecarset/sidecarset_controller.go [hunk 1]
```diff
@@ -115,9 +112,8 @@ var _ reconcile.Reconciler = &ReconcileSidecarSet{}
 // ReconcileSidecarSet reconciles a SidecarSet object
 type ReconcileSidecarSet struct {
 	client.Client
-	scheme             *runtime.Scheme
-	updateExpectations expectations.UpdateExpectations
```
Comment (zmberg) [inline pkg/controller/sidecarset/sidecarset_controller.go:119]: Why did you remove this filed? My understanding of the previous bug had nothing to do with this field being defined here, right?

Comment (wangwu50) [reply pkg/controller/sidecarset/sidecarset_controller.go:119]: Only updateExpectations were removed because they were initialized in utils

Comment (wangwu50) [reply pkg/controller/sidecarset/sidecarset_controller.go:119]: A utils is added so that the sidecarset handler can access it, similar to how cloneset handles it

Comment (zmberg) [reply pkg/controller/sidecarset/sidecarset_controller.go:119]: you can move the function to file https://github.com/openkruise/kruise/blob/master/pkg/control/sidecarcontrol/util.go.

Comment (wangwu50) [reply pkg/controller/sidecarset/sidecarset_controller.go:119]: OK, I've moved it

File: pkg/controller/sidecarset/sidecarset_pod_event_handler.go [hunk 2]
```diff
@@ -82,6 +102,9 @@ func (p *enqueueRequestForPod) updatePod(q workqueue.RateLimitingInterface, old,
 		if sidecarSet.Spec.UpdateStrategy.Type == appsv1alpha1.NotUpdateSidecarSetStrategyType {
 			continue
 		}
+		if newPod.DeletionTimestamp != nil {
```
Comment (zmberg) [inline pkg/controller/sidecarset/sidecarset_pod_event_handler.go:105]: I think you should not judge the deletionTimestamp, and use the ObserveUpdated function.

File: pkg/controller/sidecarset/sidecarset_pod_event_handler.go [hunk 3]
```diff
@@ -57,6 +74,9 @@ func (p *enqueueRequestForPod) addPod(q workqueue.RateLimitingInterface, obj run
 	}
 
 	for _, sidecarSet := range sidecarSets {
+		if pod.DeletionTimestamp != nil {
```
Comment (zmberg) [inline pkg/controller/sidecarset/sidecarset_pod_event_handler.go:77]: I think it is not necessary to execute the expectation function when add pod.

Comment (wangwu50) [reply pkg/controller/sidecarset/sidecarset_pod_event_handler.go:77]: on a restart of the controller manager, it's possible a new pod shows up in a state that is already pending deletion. Prevent the pod from being a creation observation.

File: pkg/controller/sidecarset/sidecarset_pod_event_handler.go (hunk 2 above)
Comment (wangwu50) [reply pkg/controller/sidecarset/sidecarset_pod_event_handler.go:105]: ObserveUpdated does not determine if the pod is deleted if the pod is pending deletion

Comment (wangwu50) [reply pkg/controller/sidecarset/sidecarset_pod_event_handler.go:105]: If this pod is an older version, there may be problems

**Review by furykerry — COMMENTED** [review]
/lgtm

File: pkg/controller/sidecarset/sidecarset_pod_event_handler.go [hunk 4]
```diff
@@ -82,6 +99,7 @@ func (p *enqueueRequestForPod) updatePod(q workqueue.RateLimitingInterface, old,
 		if sidecarSet.Spec.UpdateStrategy.Type == appsv1alpha1.NotUpdateSidecarSetStrategyType {
 			continue
 		}
+		sidecarcontrol.UpdateExpectations.ObserveUpdated(sidecarSet.Name, sidecarcontrol.GetSidecarSetRevision(sidecarSet), newPod)
```
Comment (veophi) [inline pkg/controller/sidecarset/sidecarset_pod_event_handler.go:102]: A little suggestion: cloud you move this Line to Line 99?

Comment (wangwu50) [reply pkg/controller/sidecarset/sidecarset_pod_event_handler.go:102]: If I put it here, I can filter out Pods managed by sidecar of type NotUpdate, right?
